File: src/engine/Divert.ts
import { Path } from "./Path";
import { PushPopType } from "./PushPop";
import { StringBuilder } from "./StringBuilder";
import { InkObject } from "./Object";
import { Pointer } from "./Pointer";
import { Container } from "./Container";
import { throwNullException } from "./NullException";

export class Divert extends InkObject {
  get targetPath() {
    if (this._targetPath != null && this._targetPath.isRelative) {
      let targetObj = this.targetPointer.Resolve();
      if (targetObj) {
        this._targetPath = targetObj.path;
      }
    }

    return this._targetPath;
  }
  set targetPath(value: Path | null) {
    this._targetPath = value;
    this._targetPointer = Pointer.Null;
  }

  public _targetPath: Path | null = null;

  get targetPointer() {
    if (this._targetPointer.isNull) {
      let targetObj = this.ResolvePath(this._targetPath).obj;

      if (this._targetPath === null)
        return throwNullException("this._targetPath");
      if (this._targetPath.lastComponent === null)
        return throwNullException("this._targetPath.lastComponent");

      if (this._targetPath.lastComponent.isIndex) {
        if (targetObj === null) return throwNullException("targetObj");
        this._targetPointer.container =
          targetObj.parent instanceof Container ? targetObj.parent : null;
        this._targetPointer.index = this._targetPath.lastComponent.index;
      } else {
        this._targetPointer = Pointer.StartOf(
          targetObj instanceof Container ? targetObj : null
        );
      }
    }

    return this._targetPointer.copy();
  }

  public _targetPointer: Pointer = Pointer.Null;

  get targetPathString() {
    if (this.targetPath == null) return null;

    return this.CompactPathString(this.targetPath);
  }
  set targetPathString(value: string | null) {
    if (value == null) {
      this.targetPath = null;
    } else {
      this.targetPath = new Path(value);
    }
  }

  public variableDivertName: string | null = null;
  get hasVariableTarget() {
    return this.variableDivertName != null;
  }

  public pushesToStack: boolean = false;
  public stackPushType: PushPopType = 0;

  public isExternal: boolean = false;
  public externalArgs: number = 0;

  public isConditional: boolean = false;

  constructor(stackPushType?: PushPopType) {
    super();
    this.pushesToStack = false;

    if (typeof stackPushType !== "undefined") {
      this.pushesToStack = true;
      this.stackPushType = stackPushType;
    }
  }

  public Equals(obj: Divert | null) {
    let otherDivert = obj;
    if (otherDivert instanceof Divert) {
      if (this.hasVariableTarget == otherDivert.hasVariableTarget) {
        if (this.hasVariableTarget) {
          return this.variableDivertName == otherDivert.variableDivertName;
        } else {
          if (this.targetPath === null)
            return throwNullException("this.targetPath");
          return this.targetPath.Equals(otherDivert.targetPath);
        }
      }
    }
    return false;
  }

  public toString() {
    if (this.hasVariableTarget) {
      return "Divert(variable: " + this.variableDivertName + ")";
    } else if (this.targetPath == null) {
      return "Divert(null)";
    } else {
      let sb = new StringBuilder();

      let targetStr = this.targetPath.toString();
      // int? targetLineNum = DebugLineNumberOfPath (targetPath);
      let targetLineNum = null;
      if (targetLineNum != null) {
        targetStr = "line " + targetLineNum;
      }

      sb.Append("Divert");

      if (this.isConditional) sb.Append("?");

      if (this.pushesToStack) {
        if (this.stackPushType == PushPopType.Function) {
          sb.Append(" function");
        } else {
          sb.Append(" tunnel");
        }
      }

      sb.Append(" -> ");
      sb.Append(this.targetPathString);

      sb.Append(" (");
      sb.Append(targetStr);
      sb.Append(")");

      return sb.toString();
    }
  }
}

File: ink_dart/lib/src/engine/divert.dart
import 'ink_object.dart';
import 'container.dart';
import 'path.dart';
import 'pointer.dart';

/// The type of push/pop operation for diverts.
enum PushPopType {
  function,
  tunnel,
  functionEvaluationFromGame,
}

/// A divert represents a jump to another location in the story.
///
/// Diverts can be:
/// - Simple jumps (->)
/// - Tunnels (->->)
/// - Function calls
/// - Conditional diverts
/// - Variable target diverts
class Divert extends InkObject {
  Path? _targetPath;
  Pointer? _targetPointer;

  /// The name of the variable containing the target (for variable diverts).
  String? variableDivertName;

  /// Whether this divert targets a variable.
  bool get hasVariableTarget => variableDivertName != null;

  /// Whether this divert pushes to the call stack.
  bool pushesToStack = false;

  /// The type of stack push (function or tunnel).
  PushPopType? stackPushType;

  /// Whether this is an external function call.
  bool isExternal = false;

  /// Number of arguments for external function calls.
  int externalArgs = 0;

  /// Whether this is a conditional divert.
  bool isConditional = false;

  /// The target path.
  Path? get targetPath {
    if (_targetPath != null && _targetPath!.isRelative) {
      final targetObj = targetPointer.element;
      if (targetObj != null) {
        _targetPath = targetObj.path;
      }
    }
    return _targetPath;
  }

  set targetPath(Path? value) {
    _targetPath = value;
    _targetPointer = null;
  }

  /// The target pointer (resolved from path).
  Pointer get targetPointer {
    if (_targetPointer == null) {
      final targetObj = resolvePath(_targetPath!).obj;
      if (targetObj == null) {
        _targetPointer = Pointer.null_();
      } else if (targetObj is Container) {
        _targetPointer = Pointer.startOf(targetObj);
      } else {
        final parentContainer = targetObj.parent as Container?;
        if (parentContainer != null) {
          _targetPointer = Pointer(
            parentContainer,
            parentContainer.content.indexOf(targetObj),
          );
        } else {
          _targetPointer = Pointer.null_();
        }
      }
    }
    return _targetPointer!;
  }

  /// The target path as a string.
  String? get targetPathString {
    if (targetPath == null) return null;
    return compactPathString(targetPath!);
  }

  set targetPathString(String? value) {
    if (value == null) {
      targetPath = null;
    } else {
      targetPath = Path.fromString(value);
    }
  }

  /// The target content object.
  InkObject? get targetContent {
    if (targetPath == null) return null;
    return resolvePath(targetPath!).obj;
  }

  @override
  String toString() {
    if (hasVariableTarget) {
      return 'Divert(variable: $variableDivertName)';
    } else if (targetPath != null) {
      final target = targetPathString ?? targetPath.toString();
      if (isConditional) {
        return 'Divert($target, conditional)';
      }
      if (pushesToStack) {
        return 'Divert($target, ${stackPushType?.name})';
      }
      return 'Divert($target)';
    }
    return 'Divert(?)';
  }

  @override
  bool operator ==(Object other) {
    if (other is Divert) {
      if (hasVariableTarget != other.hasVariableTarget) return false;
      if (hasVariableTarget) {
        return variableDivertName == other.variableDivertName;
      }
      return targetPath == other.targetPath;
    }
    return false;
  }

  @override
  int get hashCode => hasVariableTarget
      ? variableDivertName.hashCode
      : (targetPath?.hashCode ?? 0);
}
